grpc/src/builder.rs

```diff
@@ -0,0 +1,188 @@
+use http_body::Body;
+use k256::ecdsa::VerifyingKey;
+use signature::Keypair;
+use tonic::body::BoxBody;
+use tonic::client::GrpcService;
+#[cfg(not(target_arch = "wasm32"))]
+use tonic::transport::{Channel, ClientTlsConfig, Endpoint};
+
+use crate::grpc::StdError;
+use crate::{DocSigner, GrpcClient, Result, TxClient};
+
+#[cfg(not(target_arch = "wasm32"))]
+#[derive(thiserror::Error, Debug)]
+pub enum GrpcClientBuilderCertError {
+    /// Error handling certificate root
+    #[error(transparent)]
+    WebpkiError(#[from] webpki::Error),
+
+    /// Could not import system certificates
+    #[error("Could not import platform certificates: {errors:?}")]
+    RustlsNativeCertsError {
+        errors: Vec<rustls_native_certs::Error>,
+    },
+}
+
+/// Builder for [`GrpcClient`] and [`TxClient`]
+#[derive(Clone)]
+pub struct GrpcClientBuilder<T, S> {
+    connection: T,
+    signer: S,
+    account_pubkey: Option<VerifyingKey>,
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl GrpcClientBuilder<(Endpoint, ClientTlsConfig), ()> {
+    /// Create a new client connected to the given `url` using [`Channel`] transport
+    ///
+    /// [`Channel`]: tonic::transport::Channel
+    pub fn with_url(url: impl Into<String>) -> Result<Self, tonic::transport::Error> {
+        let endpoint = Endpoint::from_shared(url.into())?.user_agent("celestia-grpc")?;
+        let tls_config = ClientTlsConfig::new();
+
+        Ok(GrpcClientBuilder {
+            connection: (endpoint, tls_config),
+            signer: (),
+            account_pubkey: None,
+        })
+    }
+
+    /// Enables the platform’s trusted certs.
+    #[cfg(feature = "tls-native-roots")]
+    pub fn with_native_roots(self) -> Result<Self, GrpcClientBuilderCertError> {
+        let rustls_native_certs::CertificateResult { certs, errors, .. } =
+            rustls_native_certs::load_native_certs();
+
+        if certs.is_empty() {
+            return Err(errors.into());
+        }
+        let (endpoint, tls_config) = self.connection;
+
+        let tls_config = tls_config.trust_anchors(
+            certs
+                .into_iter()
+                .map(|c| Ok(webpki::anchor_from_trusted_cert(&c)?.to_owned()))
+                .collect::<Result<Vec<_>, GrpcClientBuilderCertError>>()?,
+        );
+
+        Ok(Self {
+            connection: (endpoint, tls_config),
+            ..self
+        })
+    }
+
+    /// Enables the webpki roots.
+    #[cfg(feature = "tls-webpki-roots")]
+    pub fn with_webpki_roots(self) -> Self {
+        let roots = webpki_roots::TLS_SERVER_ROOTS.iter().cloned();
+        let (endpoint, tls_config) = self.connection;
+        Self {
+            connection: (endpoint, tls_config.trust_anchors(roots)),
+            ..self
+        }
+    }
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl<S> GrpcClientBuilder<(Endpoint, ClientTlsConfig), S> {
+    /// Connects to the endpoint, applying tls configuration
+    pub fn connect(self) -> Result<GrpcClientBuilder<Channel, S>> {
```

Comment:
I don't like having to call both connect and build, can we handle that in build?